web/client/src/models/sqlmesh-model.ts

```diff
@@ -0,0 +1,158 @@
+    this.lineage = this.initial.lineage ?? {}
+  }
+
+  update(initial: Partial<InitialSQLMeshModel> = {}): void {
+    for (const [key, value] of Object.entries(initial)) {
+      if (key === 'columns') {
+        this.columns = value as Column[]
+      } else if (key === 'details') {
+        this.details = value as ModelDetails
+      } else if (key === 'lineage') {
+        this.lineage = value as Record<string, Lineage>
+      } else if (key in this) {
+        this[key as 'path' | 'name' | 'dialect' | 'description' | 'sql'] =
+          value as string
+      }
+    }
+  }
+
+  // TODO: use better merge
+  static mergeLineage(
+    models: Map<string, Model>,
+    lineage: Record<string, Lineage>,
+    columns: ColumnLineageApiLineageModelNameColumnNameGet200 = {},
+  ): Record<string, Lineage> {
+    lineage = structuredClone(lineage)
+
+    for (const model in columns) {
+      const lineageModel = lineage[model]
+      const columnsModel = columns[model]
+
+      if (lineageModel == null || columnsModel == null) continue
+
+      if (lineageModel.columns == null) {
+        lineageModel.columns = {}
+      }
+
+      for (const columnName in columnsModel) {
+        const columnsModelColumn = columnsModel[columnName]
+
+        if (columnsModelColumn == null) continue
+
+        const lineageModelColumn = lineageModel.columns[columnName] ?? {}
+
+        lineageModelColumn.source = columnsModelColumn.source
+        lineageModelColumn.models = {}
+
+        lineageModel.columns[columnName] = lineageModelColumn
+
+        if (isObjectEmpty(columnsModelColumn.models)) continue
+
+        for (const columnModel in columnsModelColumn.models) {
+          const columnsModelColumnModel = columnsModelColumn.models[columnModel]
+
+          if (columnsModelColumnModel == null) continue
+
+          const lineageModelColumnModel = lineageModelColumn.models[columnModel]
+
+          if (lineageModelColumnModel == null) {
+            lineageModelColumn.models[columnModel] = columnsModelColumnModel
+          } else {
+            lineageModelColumn.models[columnModel] = Array.from(
+              new Set(lineageModelColumnModel.concat(columnsModelColumnModel)),
+            )
+          }
+        }
+      }
+    }
+
+    for (const modelName in lineage) {
+      const model = models.get(modelName)
+      const modelLineage = lineage[modelName]
+
+      if (model == null || modelLineage == null) {
+        delete lineage[modelName]
+
+        continue
+      }
+
+      if (modelLineage.columns == null) continue
+
+      if (model.columns == null) {
+        delete modelLineage.columns
+
+        continue
+      }
+
+      for (const columnName in modelLineage.columns) {
+        const found = model.columns.find(c => c.name === columnName)
+
+        if (found == null) {
+          delete modelLineage.columns[columnName]
+
+          continue
+        }
+      }
+    }
+
+    return lineage
+  }
+
+  static encodeName(modelName: string): string {
```

Comment:
Why is this needed?

Reply:
if we do `/docs/models/sushi.customers` browser thinks `.customers` is top-level domain i guess, so i'm juts replacing `.`